### PR TITLE
parts: replace snapcraft command-chain with environment

### PR DESCRIPTION
**Pull request by cmatsuoka**
Changes for core22 snaps:
1. Pass `command_chain` to snap.yaml.
2. Deprecate `adapter` field.
3. Replace `snapcraft_runner` and it's entry in the `command_chain` with global environment variables (`LD_LIBRARY_PATH` and `PATH`).
4. Assume `command-chain` if app or hook `command-chain`s are declared.

<br />

Behavior for overriding `LD_LIBRARY_PATH` and `PATH` in snapcraft.yaml:
- If `LD_LIBRARY_PATH` or `PATH` are defined, keep user-defined values.
- If `LD_LIBRARY_PATH` or `PATH` are not defined, set to default values.
- If `LD_LIBRARY_PATH` or `PATH` are null, do not use default values

<br />

Default values: 
- `LD_LIBRARY_PATH`: `$SNAP_LIBRARY_PATH:$LD_LIBRARY_PATH`
- `PATH`: `$SNAP/usr/sbin:$SNAP/usr/bin:$SNAP/sbin:$SNAP/bin:$PATH`

<br />

LP: #[1973173](https://bugs.launchpad.net/snapcraft/+bug/1973173)
CRAFT-1065

Signed-off-by: Callahan Kovacs <callahan.kovacs@canonical.com>
Co-authored-by: Claudio Matsuoka <claudio.matsuoka@canonical.com>

- [x] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [x] Have you successfully run `make lint`?
- [x] Have you successfully run `pytest tests/unit`?

-----

### CODE REVIEW

File: snapcraft/parts/setup_assets.py [hunk 1]
```diff
@@ -46,12 +46,21 @@ def setup_assets(project: Project, *, assets_dir: Path, prime_dir: Path) -> None
...
+            if hook.command_chain:
+                _validate_command_chain(
+                    hook.command_chain, name=f"hook {hook_name!r}", prime_dir=prime_dir
+                )
 
     if not project.apps:
         return
 
+    # If there are no apps, or type is snapd, no need to create a runner.
+    if project.type != "snapd":
```
[the remark this answers is not in the document]
Comment (cmatsuoka) [reply snapcraft/parts/setup_assets.py:61]: "bare", and also no base?

File: snapcraft/meta/snap_yaml.py [hunk 2]
```diff
@@ -155,11 +160,15 @@ def write(project: Project, prime_dir: Path, *, arch: str):
...
-                adapter=app.adapter,
-                command_chain=["snap/command-chain/snapcraft-runner"],
+                command_chain=command_chain if command_chain else None,
                 sockets=app_sockets if app_sockets else None,
             )
 
+    if project.hooks:
+        for hook in project.hooks.values():
+            if hook.command_chain:
+                assumes.add("command-chain")
```
Comment (sergiusens) [inline snapcraft/meta/snap_yaml.py:170]:
```suggestion
        if any(h for h in in project.hooks.values() if h.command_chain):
            assumes.add("command-chain")
```

File: snapcraft/meta/snap_yaml.py [hunk 3]
```diff
@@ -131,6 +132,10 @@ def write(project: Project, prime_dir: Path, *, arch: str):
                         socket_mode=socket.socket_mode,
                     )
 
+            command_chain = app.command_chain
```
Comment (sergiusens) [inline snapcraft/meta/snap_yaml.py:135]: this seems unnecessary

File: snapcraft/parts/setup_assets.py [hunk 4]
```diff
@@ -49,8 +49,13 @@ def setup_assets(
     gui_dir.mkdir(parents=True, exist_ok=True)
 
     _write_snap_directory(assets_dir=assets_dir, prime_dir=prime_dir, meta_dir=meta_dir)
-    _write_snapcraft_runner(prime_dir=prime_dir)
-    # TODO: write snapcraft
```
Comment (sergiusens) [inline snapcraft/parts/setup_assets.py:53]: why is this TODO gone? Are we writing snapcraft.yaml and manifest.yaml now?

File: snapcraft/parts/setup_assets.py [hunk 5]
```diff
@@ -77,6 +86,10 @@ def setup_assets(
         relative_icon_path = str(icon_path)
 
     for app_name, app in project.apps.items():
+        _validate_command_chain(
+            app.command_chain, name=f"app {app_name!r}", prime_dir=prime_dir
+        )
+
         if not app.desktop:
             continue
```
Comment (sergiusens) [inline snapcraft/parts/setup_assets.py:94]: this continue in the middle has proved probalematic, I would change this now and do `if app.desktop` and nest the desktop file logic below

File: snapcraft/parts/setup_assets.py (hunk 4 above)
Comment (mr-cal) [reply snapcraft/parts/setup_assets.py:53]: I believe it is a duplicate of the TODO on line 237:
`# TODO: record manifest and source snapcraft.yaml`

**Comment by mr-cal** [comment]
Failure for flutter spread test and the linting appear to be unrelated.

File: snapcraft/projects.py [hunk 6]
```diff
@@ -270,7 +269,7 @@ class Project(ProjectModel):
     epoch: Optional[str]
     adopt_info: Optional[str]
     system_usernames: Optional[Dict[str, Any]]
-    environment: Optional[Dict[str, str]]
+    environment: Optional[Dict[str, Any]]
```
Comment (sergiusens) [inline snapcraft/projects.py:272]: this might be better kept as a str as environment variable values are just that, strings

Comment (mr-cal) [reply snapcraft/projects.py]: Good catch.

File: tests/spread/general/unicode-metadata/task.yaml [hunk 7]
```diff
@@ -10,6 +10,10 @@ prepare: |
 
   base=$(get_base)
   sed -e "s/base: {{BASE}}/base: ${base}/g" expected_snap_tmpl.yaml > expected_snap.yaml
+  if [[ "$base" =~ "core22" ]]; then
```
Comment (mr-cal) [inline tests/spread/general/unicode-metadata/task.yaml:13]: I couldn't figure out a better way to make this spread test work across core18, 20, and 22.

Comment (sergiusens) [reply tests/spread/general/unicode-metadata/task.yaml:13]: this is fine for now, we will eventually hit a problem as we are not resetting the expected_snap.yaml, but I would not worry about that now.

**Review by sergiusens — APPROVED** [review]
Nice work, also kudos on the extra effort on command chain testing

**Pull request merged**